File: cef-web/src/main/webapp/src/app/core/services/lookup.service.ts
/*
 * © Copyright 2019 EPA CAERS Project Team
 *
 * This file is part of the Common Air Emissions Reporting System (CAERS).
 *
 * CAERS is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * CAERS is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with CAERS.  If
 * not, see <https://www.gnu.org/licenses/>.
*/
import { Injectable } from '@angular/core';
import { BaseCodeLookup } from 'src/app/shared/models/base-code-lookup';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { Pollutant } from 'src/app/shared/models/pollutant';
import { CalculationMethodCode } from 'src/app/shared/models/calculation-method-code';
import { UnitMeasureCode } from 'src/app/shared/models/unit-measure-code';
import { FipsStateCode } from 'src/app/shared/models/fips-state-code';
import { FacilityNaicsCode } from 'src/app/shared/models/facility-naics-code';
import { AircraftEngineTypeCode } from 'src/app/shared/models/aircraft-engine-type-code';
import { PointSourceSccCode } from 'src/app/shared/models/point-source-scc-code';
import { FacilityCategoryCode } from 'src/app/shared/models/facility-category-code';
import { FipsCounty } from 'src/app/shared/models/fips-county';
import { InventoryYearCodeLookup } from 'src/app/shared/models/inventory-year-code-lookup';
import { CalculationMaterialCode } from 'src/app/shared/models/calculation-material-code';
import { MasterFacilityNaicsCode } from 'src/app/shared/models/master-facility-naics-code';
import {PermitType} from "src/app/shared/models/permit-type";

@Injectable({
  providedIn: 'root'
})
export class LookupService {

  private baseUrl = 'api/lookup';  // URL to web api

  constructor(private http: HttpClient) { }

  retrieveCalcMaterial(): Observable<CalculationMaterialCode[]> {
    const url = `${this.baseUrl}/calculation/material`;
    return this.http.get<CalculationMaterialCode[]>(url);
  }

  retrieveFuelUseMaterial(): Observable<CalculationMaterialCode[]> {
    const url = `${this.baseUrl}/fuelUse/material`;
    return this.http.get<CalculationMaterialCode[]>(url);
  }

	retrieveMonthlyReportingMaterial(): Observable<CalculationMaterialCode[]> {
    const url = `${this.baseUrl}/monthlyReporting/material`;
    return this.http.get<CalculationMaterialCode[]>(url);
  }

  retrieveCalcMethod(): Observable<CalculationMethodCode[]> {
    const url = `${this.baseUrl}/calculation/method`;
    return this.http.get<CalculationMethodCode[]>(url);
  }

  retrieveCalcParam(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/calculation/parameter`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveSubFacilityOperatingStatus(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/subFacilityOperatingStatus`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveFacilityOperatingStatus(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/facilityOperatingStatus`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveEmissionsOperatingType(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/emissionsOperatingType`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrievePollutant(): Observable<Pollutant[]> {
    const url = `${this.baseUrl}/pollutant`;
    return this.http.get<Pollutant[]>(url);
  }

  retrieveCurrentPollutants(year: number): Observable<Pollutant[]> {
    const url = `${this.baseUrl}/pollutant/${year}`;
    return this.http.get<Pollutant[]>(url);
  }

  retrieveReportingPeriod(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/reportingPeriod`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveUom(year: number): Observable<UnitMeasureCode[]> {
    const url = `${this.baseUrl}/uom/${year}`;
    return this.http.get<UnitMeasureCode[]>(url);
  }

  retrieveFuelUseUom(): Observable<UnitMeasureCode[]> {
    const url = `${this.baseUrl}/fuelUse/uom`;
    return this.http.get<UnitMeasureCode[]>(url);
  }

  retrieveFacilityContactType(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/contactType`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveFipsCounties(): Observable<FipsCounty[]> {
    const url = `${this.baseUrl}/county`;
    return this.http.get<FipsCounty[]>(url);
  }

  retrieveCurrentFipsCounties(year: number): Observable<FipsCounty[]> {
    const url = `${this.baseUrl}/county/${year}`;
    return this.http.get<FipsCounty[]>(url);
  }

  retrieveFipsCountiesForState(stateCode: string): Observable<FipsCounty[]> {
    const url = `${this.baseUrl}/county/state/${stateCode}`;
    return this.http.get<FipsCounty[]>(url);
  }

  retrieveCurrentFipsCountiesForState(stateCode: string, year: number): Observable<FipsCounty[]> {
    const url = `${this.baseUrl}/county/state/${stateCode}/${year}`;
    return this.http.get<FipsCounty[]>(url);
  }

  retrieveFipsStateCode(): Observable<FipsStateCode[]> {
    const url = `${this.baseUrl}/stateCode`;
    return this.http.get<FipsStateCode[]>(url);
  }

  retrieveUnitType(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/unitType`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveReleaseTypeCode(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/releaseType`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveCurrentReleaseTypeCodes(year: number): Observable<InventoryYearCodeLookup[]> {
    const url = `${this.baseUrl}/releasePointType/${year}`;
    return this.http.get<InventoryYearCodeLookup[]>(url);
  }

  retrieveProgramSystemTypeCode(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/programSystemType`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveProgramSystemCodeByDescription(description: string): Observable<BaseCodeLookup> {
    const url = `${this.baseUrl}/programSystem/description/${description}`;
    return this.http.get<BaseCodeLookup>(url);
  }

  retrieveControlMeasureCodes(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/controlMeasure`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveCurrentControlMeasureCodes(year: number): Observable<InventoryYearCodeLookup[]> {
    const url = `${this.baseUrl}/controlMeasure/${year}`;
    return this.http.get<InventoryYearCodeLookup[]>(url);
  }

  retrieveTribalCode(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/tribalCode`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveNaicsCode(): Observable<FacilityNaicsCode[]> {
    const url = `${this.baseUrl}/naicsCode`;
    return this.http.get<FacilityNaicsCode[]>(url);
  }

  retrieveNaicsCodesByYear(year: number): Observable<FacilityNaicsCode[]> {
    const url = `${this.baseUrl}/naicsCode/${year}`;
    return this.http.get<FacilityNaicsCode[]>(url);
  }

  retrieveCurrentMFNaicsCodes(year: number): Observable<MasterFacilityNaicsCode[]> {
    const url = `${this.baseUrl}/naicsCode/${year}`;
    return this.http.get<MasterFacilityNaicsCode[]>(url);
  }

  retrievePermitTypes(): Observable<PermitType[]> {
    const url = `${this.baseUrl}/permitType`;
    return this.http.get<PermitType[]>(url);
  }

  retrieveAircraftEngineCodes(scc: string): Observable<AircraftEngineTypeCode[]> {
    const url = `${this.baseUrl}/aircraftEngineCode/${scc}`;
    return this.http.get<AircraftEngineTypeCode[]>(url);
  }

  retrieveCurrentAircraftEngineCodes(scc: string, year: number): Observable<AircraftEngineTypeCode[]> {
    const url = `${this.baseUrl}/aircraftEngineCode/${scc}/${year}`;
    return this.http.get<AircraftEngineTypeCode[]>(url);
  }

  retrievePointSourceSccCode(code: string): Observable<PointSourceSccCode> {
    const url = `${this.baseUrl}/pointSourceSccCode/${code}`;
    return this.http.get<PointSourceSccCode>(url);
  }

  retrieveFacilityCategory(): Observable<FacilityCategoryCode[]> {
    const url = `${this.baseUrl}/facility/category`;
    return this.http.get<FacilityCategoryCode[]>(url);
  }

  retrieveFacilitySourceType(): Observable<BaseCodeLookup[]> {
    const url = `${this.baseUrl}/facility/sourceType`;
    return this.http.get<BaseCodeLookup[]>(url);
  }

  retrieveCurrentFacilitySourceType(year: number): Observable<InventoryYearCodeLookup[]> {
    const url = `${this.baseUrl}/facility/sourceType/${year}`;
    return this.http.get<InventoryYearCodeLookup[]>(url);
  }

  basicSccSearch(searchTerm: string, slt: string): Observable<PointSourceSccCode[]> {
	const url = `${this.baseUrl}/scc/${slt}/${searchTerm}`;
	return this.http.get<PointSourceSccCode[]>(url);
  }

}

File: cef-web/src/main/java/gov/epa/cef/web/service/LookupService.java
/*
 * © Copyright 2019 EPA CAERS Project Team
 *
 * This file is part of the Common Air Emissions Reporting System (CAERS).
 *
 * CAERS is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * CAERS is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with CAERS.  If
 * not, see <https://www.gnu.org/licenses/>.
*/
package gov.epa.cef.web.service;

import gov.epa.cef.web.domain.*;
import gov.epa.cef.web.service.dto.*;

import java.util.List;

public interface LookupService {

    /**
     * Retrieve Calculation Material codes
     * @return
     */
    List<CodeLookupDto> retrieveCalcMaterialCodes();

    /**
     * Retrieve Fuel Use and Asphalt Calculation Material codes
     * @return
     */
    List<CalculationMaterialCodeDto> retrieveFuelUseMaterialCodes();

    /**
     * Retrieves Fuel Use Material Code entity by code
     * @param code
     * @return CalculationMaterialCode
     * */
    CalculationMaterialCode retrieveFuelUseMaterialCodeEntityByCode(String code);

    /**
     * Retrieve Calculation Material code database object by code
     * @param code
     * @return
     */
    CalculationMaterialCode retrieveCalcMaterialCodeEntityByCode(String code);

    /**
     * Retrieve Calculation Method codes
     * @return
     */
    List<CalculationMethodCodeDto> retrieveCalcMethodCodes();

    /**
     * Retrieve Calculation Method Code entity by code
     * @param code
     * @return CalculationMethodCode
     */
    CalculationMethodCode retrieveCalcMethodCodeEntityByCode(String code);

    /**
     * Retrieve Calculation Parameter Type codes
     * @return
     */
    List<CodeLookupDto> retrieveCalcParamTypeCodes();

    /**
     * Retrieve Calculation Parameter Type code database object by code
     * @param code
     * @return
     */
    CalculationParameterTypeCode retrieveCalcParamTypeCodeEntityByCode(String code);

    /**
     * Retrieve Operating Status codes for sub-facility components
     * @return
     */
    List<CodeLookupDto> retrieveSubFacilityOperatingStatusCodes();

    /**
     * Retrieve OperatingStatusCode entities for sub-facility components by code
     * @param code;
     * @return OperatingStatusCode
     */
    OperatingStatusCode retrieveSubFacilityOperatingStatusCodeEntityByCode(String code);

    /**
     * Retrieve Operating Status codes for facilities
     * @return
     */
    List<CodeLookupDto> retrieveFacilityOperatingStatusCodes();

    /**
     * Retrieve Pollutants
     * @return
     */
    List<PollutantDto> retrievePollutants();

    /**
     * Retrieve non-legacy Pollutants
     * @param year
     * @return
     */
    List<PollutantDto> retrieveCurrentPollutants(Integer year);

    /**
     * Retrieve Pollutant by Pollutant Code
     * @param pollutantCode
     * @return Pollutant
     */
    Pollutant retrievePollutantByPollutantCode(String pollutantCode);

    /**
     * Retrieve Operating Status code database object by code
     * @param code
     * @return
     */
    OperatingStatusCode retrieveOperatingStatusCodeEntityByCode(String code);

    /**
     * Retrieve Reporting Period codes
     * @return
     */
    List<CodeLookupDto> retrieveReportingPeriodCodes();

    /**
     * Retrieve Reporting Period code database object by code
     * @param code
     * @return
     */
    ReportingPeriodCode retrieveReportingPeriodCodeEntityByCode(String code);

    /**
     * Retrieve UoM codes
     * @return
     */
    List<UnitMeasureCodeDto> retrieveUnitMeasureCodes();

    /**
     * Retrieve non-legacy UoM codes
     * @return
     */
    List<UnitMeasureCodeDto> retrieveCurrentUnitMeasureCodes(Integer year);

    /**
     * Retrieve Fuel Use UoM codes
     * @return
     */
    List<UnitMeasureCodeDto> retrieveFuelUseUnitMeasureCodes();

    /**
     * Retrieve UoM code database object by code
     * @param code
     * @return
     */
    UnitMeasureCode retrieveUnitMeasureCodeEntityByCode(String code);

    /**
     * Retrieve the list of Emission Operating Type Codes
     * @return
     */
    List<CodeLookupDto> retrieveEmissionOperatingTypeCodes();

    /**
     * Retrieve the Emissions Operating Type Code entity by code
     *
     * @param code
     * @return
     */
    EmissionsOperatingTypeCode retrieveEmissionsOperatingTypeCodeEntityByCode(String code);

    /**
     * Retrieve Contact Type codes
     * @return
     */
    List<CodeLookupDto> retrieveContactTypeCodes();

    /**
     * Retrieve Contact Type code database object by code
     * @param code
     * @return
     */
    ContactTypeCode retrieveContactTypeEntityByCode(String code);
    /**
     * Retrieve Unit Type codes
     * @return
     */
    List<CodeLookupDto> retrieveUnitTypeCodes();

    /**
     * Retrieve UnitTypeCode entity by typeCode
     * @param typeCode
     * @return UnitTypeCode
     * */
    UnitTypeCode retrieveUnitTypeCodeEntityByCode(String typeCode);

    /**
     * Retrieve County codes
     * @return
     */
    List<FipsCountyDto> retrieveCountyCodes();

    /**
     * Retrieve non-legacy County codes
     * @param year
     * @return
     */
    List<FipsCountyDto> retrieveCurrentCounties(Integer year);

    /**
     * Retrieve County codes for a state
     * @param stateCode
     * @return
     */
    List<FipsCountyDto> retrieveCountyCodesByState(String stateCode);

    /**
     * Retrieve non-legacy County codes for a state
     * @param stateCode
     * @param year
     * @return
     */
    List<FipsCountyDto> retrieveCurrentCountyCodesByState(String stateCode, Integer year);


    /**
     * Retrieve FipsCounty entity by code
     * @param stateCode
     * @param countyCode
     * @return
     */
    FipsCounty retrieveCountyEntityByStateAndCountyCode(String stateCode, String countyCode);

    /**
     * Retrieve FipsCounty entity by countyCode
     * @param countyCode
     * @return
     */
    FipsCounty retrieveCountyEntityByCountyCode(String countyCode);

    /**
     * Retrieve Fips State codes
     * @return
     */
    List<FipsStateCodeDto> retrieveStateCodes();

    /**
     * Retrieve Fips State code database object by code
     * @param code
     * @return
     */
    FipsStateCode retrieveStateCodeEntityByCode(String code);

    /**
     * Retrieve Fips State code database object by uspsCode
     * @param uspsCode
     * @return
     */
    FipsStateCode retrieveStateCodeEntityByUspsCode(String uspsCode);

    /**
     * Retrieve the list of Release Point Type Codes
     * @return
     */
    List<CodeLookupDto> retrieveReleasePointTypeCodes();

    /**
     * Retrieve Release Point Type code database object by code
     * @param code
     * @return
     */
    ReleasePointTypeCode retrieveReleasePointTypeCodeEntityByCode(String code);

    /**
     * Retrieve non-legacy Release Point Type codes
     * @param year
     * @return
     */
    List<CodeLookupDto> retrieveCurrentReleasePointTypeCodes(Integer year);

    /**
     * Retrieve the list of Program System Type Codes
     * @return
     */
    List<CodeLookupDto> retrieveProgramSystemTypeCodes();

    /**
     * Retrieve Program System code by description
     * @param description
     * @return
     */
    CodeLookupDto retrieveProgramSystemCodeByDescription(String description);

    /**
     * Retrieve Program System Type code database object by code
     * @param code
     * @return
     */
    ProgramSystemCode retrieveProgramSystemTypeCodeEntityByCode(String code);

    /**
     * Retrieve Control Measure codes
     * @return
     */
    List<CodeLookupDto> retrieveControlMeasureCodes();

    /**
     * Retrieve non-legacy Control Measure codes
     * @param year
     * @return
     */
    List<CodeLookupDto> retrieveCurrentControlMeasureCodes(Integer year);

    /**
    * Retrieve Control Measure code database object by code
    * @param code
    * @return
    */
    ControlMeasureCode retrieveControlMeasureCodeEntityByCode(String code);

    /**
     * Retrieve Tribal Codes
     * @return
     */
    List<CodeLookupDto> retrieveTribalCodes();

    /**
     * Retrieve Tribal code database object by code
     * @param code
     * @return
     */
    TribalCode retrieveTribalCodeEntityByCode(String code);

    /**
     * Retrieve NAICS Codes
     * @return List of CodeLookupDtos
     */
    List<CodeLookupDto> retrieveNaicsCodes();

    /**
     * Retrieve NAICS code entity by code
     * @param code
     * @return
     */
    NaicsCode retrieveNAICSCodeEntityByCode(Integer code);

    /**
     * Retrieve NAICS codes by year and boolean to limit results to six-digit codes, sort results by sortingField
     * @param year
     * @param onlySixDigits
     * @param sortingField
     * @return
     */
    List<CodeLookupDto> retrieveNaicsCodesByYear(Short year, boolean onlySixDigits, String sortingField);

    List<FacilityPermitTypeDto> retrieveFacilityPermitTypes();

    /**
     * Retrieve Aircraft Engine Type Codes
     * @return
     */
    List<AircraftEngineTypeCodeDto> retrieveAircraftEngineCodes(String scc);

    /**
     * Retrieve AircraftEngineTypeCode entity by code
     * @param code
     * @return AircraftEngineTypeCode
     * */
    AircraftEngineTypeCode retrieveAircraftEngineCodeEntityByCode(String code);

    /**
     * Retrieve non-legacy Aircraft Engine Type Codes
     * @param scc
     * @param year
     * @return
     */
    List<AircraftEngineTypeCodeDto> retrieveCurrentAircraftEngineCodes(String scc, Integer year);

    /**
    * Retrieve Point Source SCC code database object by code
    * @param code
    * @return
    */
    PointSourceSccCodeDto retrievePointSourceSccCode(String code);

    /**
     * Retrieve Facility Category codes
     * @return
     */
    List<FacilityCategoryCodeDto> retrieveFacilityCategoryCodes();

    /**
     * Retrieve Facility Category Code entity by code
     * @param code
     * @return FacilityCategoryCode
     */
    FacilityCategoryCode retrieveFacilityCategoryCodeEntityByCode(String code);

    /**
     * Retrieve Facility Source Type codes
     * @return
     */
    List<CodeLookupDto> retrieveFacilitySourceTypeCodes();

    /**
     * Retrieve non-legacy Facility Source Type codes
     * @param year
     * @return
     */
    List<CodeLookupDto> retrieveCurrentFacilitySourceTypeCodes(Integer year);


    /**
     * Retrieve scc codes by search term. SLT (or program system code) is provided
     * for checking what SCCs are allowed.
     * @param searchTerm
     * @param slt
     * @return
     */
    List<PointSourceSccCodeDto> searchSccCodes(String searchTerm, String slt);

    /**
     * Retrieve reporting period codes for all 12 months
     * @return
     */
    List<ReportingPeriodCode> findReportingPeriodMonthAndSemiAnnualCodes();

    /**
     * Retrieve NAICS code description per given report year
     * @param code
     * @param year
     * @return
     */
    String findNaicsCodeDescriptionByYear(Integer code, Short year);

    /**
     * Retrieve Emission Formula Variable Code entity by code
     * @param code
     * @return EmissionFormulaVariableCode
     */
    EmissionFormulaVariableCode retrieveFormulaVariableCodeEntityByCode(String code);

    /**
     * Check if a pollutant was reported in the previous year's report
     * @param pollutantCode
     * @param masterFacilityId
     * @return boolean
     */
    boolean wasPollutantReportedLastYear(String pollutantCode, Long masterFacilityId);

}
